### DISCUSSION TITLE
Elks SSH client

### DISCUSSION THREAD

**Discussion by rafael2k** [jbruchon/elks#1786] [General]
Hi there! Is there any ssh client for Elks?
I realized there is SSH2DOS, which does support real-mode and 8086 natively:
https://sshdos.sourceforge.net/history2.txt

**Comment by rafael2k** [comment]
It is an Open Wattcom project (DJGPP support removed around version 0.1), with realmode-8086 project target:
https://github.com/AnttiTakala/SSH2DOS

It uses the large memory mode for compilation.

**Reply by ghaerr** [reply]
We don't have an SSH client for ELKS. 

The SSH2DOS project looks interesting, but there's lots of work to use it for ELKS. A quick look shows the following issues which would need to be handled somehow:
- large executable: the ELKS compiler doesn't support large model (=far code/far data), only far code with a limit of 128kb. This means the programs may not actually link if they're too large, and each source file may have to be modified to include `__far` to support far pointers, along with new memory management routines for allocations and possibly far libc routines.
- 32-bit int default: some source looks like it requires 32-bit `int` as default. All cases would have to be looked at and `long` substituted instead in each source file.
- watcom TCP: the programs use a non-standard DOS-based TCP/IP instead of Berkeley sockets. Most of the lower level networking routines need to be rewritten for ELKS (Berkeley) sockets.
- testing: deep knowledge of SSH is likely required to get this all working, given the above.

**Reply by Oichkatzelesfrettschen** [reply]
My understanding is right only 12% of the time. But here are some of the issues I'm trying to imagine to expand upon your list and then just sort of dream-scribble some ideas based on "what I know." : 

- ELKS supports only far code up to 128KB.
- Would need to refactor the code to use __far pointers ... (annotate data structures and functions with __far) ... and implement new memory management routines of some kind.
- Split large functions into smaller segments if they exceed 128KB
- Replace 32-bit ints with long types in all relevant source files... search for int declarations and replace with long where 32-bit is assumed... 
- Adapt `tcp_connect `to use `socket`, `connect `from ELKS.
- Replace `tcp_send`, `tcp_receive `with `send`, `recv `calls.
- Replace `net_init `with ELKS network initialization stuffs
- Adapt net_close to clean up ELKS socket resources properly.

So `tcp_connect `would be: 

```
int sockfd = socket(AF_INET, SOCK_STREAM, 0);
struct sockaddr_in server_addr;
// Set up server_addr with IP and port
connect(sockfd, (struct sockaddr *)&server_addr, sizeof(server_addr));
```
Send and receive: 
```
int bytes_sent = send(sockfd, buffer, length, 0);
int bytes_received = recv(sockfd, buffer, length, 0);
```
Then the magical task of initializing the ELKS network stack...

`net_init(); // Ensure ELKS network stack is initialized`

And saying goodbye... 

`close(sockfd); // Close the socket using ELKS API`

Easy peasy, right? Eeeeep.

**Reply by ghaerr** [reply]
Hello @Oichkatzelesfrettschen,

Yes, you're pretty much correct overall in your assumptions when using the `ia16-elf-gcc` compiler and linker, along with the current ELKS kernel. More specifically though, the 128k code size is the limit for *all* functions combined, not just one function. Using `__far` pointers can work, but even their use is limited to an array or object that's 64k or less. Much more reengineering would be required if a program needs access to a larger object from a single far pointer.

Finally, `send` and `receive` aren't actually implemented, only `read` and `write` and only TCP, no UDP functionality or IP layer access. This probably isn't a problem, but it'd be a show-stopper if the application required special flags or UDP.

I have just learned more about the possibilities of using Watcom C for ELKS applications, which could help greatly. See my comment below.

**Comment by rafael2k** [comment]
Thanks @ghaerr ! Got it. I imagined it would be a bit more than a recompilation.
: )
Concerning the large model, would it be possible to use the Wattcom compiler to create the elks executable? I understand the libc would also need some work, as well code adaptation to match the correct semantic of the int size.
Changing watcom tcp calls to berkeley sockets would be the fun part I think. Anyway, I'll also take a look in other implementations to check if there is any better option.

**Reply by ghaerr** [reply]
Using Watcom could be a neat idea, but the `wlink` linker would have to be modified to emit an ELKS-compatible a.out format executable, rather than DOS-compatible .exe format. Other options would be a post-processor for .exe files to convert to a.out, or having ELKS load .exe files directly. All would be potentially a lot of work.

**Comment by rafael2k** [comment]
I have a feeling better I focus on compiling in gcc-ia16 and strip the wattcom tcp stack and see what I get.

**Reply by ghaerr** [reply]
If you're serious about trying to get this to work the first step would be attempting a compile/link just to see whether the code size is < 128k, as that's a hard limit. I checked the DOS .exe and its 143k so it could be close, depending on the amount of initialized data. It would be worth continuing to check around for small SSH implementations.

**Reply by ghaerr** [reply]
Hello @rafael2k,

> It is an Open Wattcom project, with realmode-8086 project target
> https://github.com/AnttiTakala/SSH2DOS

> better I focus on compiling in gcc-ia16 and strip the wattcom tcp stack and see what I get.

I have found an interesting possibility of using the OpenWatcom C compiler and linker that could, although with lots of work, allow large model (large code and data) programs to run on ELKS. The [owtarget16 Project](https://github.com/pts/owtarget16) has completed a proof of concept allowing C programs compiled with OpenWatcom to run on ELKS. Still lots to do, as there is as yet no C library for it, and large model may require kernel enhancements. [See my comment here](https://github.com/pts/owtarget16/issues/1) for more details about the potential issues.

Another very interesting project by the same author @pts [minilibc686](https://github.com/pts/minilibc686) contains a utility `omf2elf` which converts OpenWatcom (or other MS compiler) .OBJ files into ELF .o files. This could be then used in similar fashion but using the ELKS `ia16-elf-ld.gold` linker along with the ELKS C library to produce an executable compiled from Watcom C. In this case, there would still be a 128k code limit but large model far data might be more easily supported. I'm looking into both solutions.

Nonetheless, if SSH2DOS actually runs on 16-bit DOS and compiles with OpenWatcom C, that same source would probably be a good start for an ELKS port, for someone interested in really digging into it.

**Comment by rafael2k** [comment]
The code will fit in 128kB. The binary distributed is large because it includes the whole Wattcom TCP/IP stack. Even dropbear ssh server and client code compiled and linked is less than 100kB [1] on 32bit OSes.

[1] https://github.com/mkj/dropbear

**Reply by ghaerr** [reply]
I didn't know Watcom had a TCP/IP stack. Do you know if the source for that was ever released somewhere?

**Reply by rafael2k** [reply]
Source mentions "Waterloo TCP/IP":
https://github.com/open-watcom/open-watcom-v2/tree/master/bld/wattcp

**Reply by ghaerr** [reply]
Wow! The Open Watcom C source is filled with good stuff. I've been studying it lately and have managed to get the entire toolchain up and running on macOS, including its C library (although the system calls aren't compatible with ELKS which is a whole 'nother project). The compiler then builds all its libraries and I can see now that it has built them all, including TCP/IP on my system for 8086 real mode DOS. That's kind of amazing. The more I'm learning about Watcom C the more I'm realizing the amount of work that went into it and that its a truly professional compiler toolchain and set of libraries running on many platforms.

**Reply by rafael2k** [reply]
Indeed we need to take our hat off, especially for them open-sourcing all their code!
: )

**Reply by tyama501** [reply]
OpenWatcom is the main compiler of the FreeDOS kernel.
Wattcp is included in FreeDOS distribution, although mtcp may be used.

**Reply by ghaerr** [reply]
I also notice looking at OpenWatcom's C library source that the BIOS support routines for IBM PC are also supported specially for PC-98. There appears to be a Japanese GUI version of OWC as well. It appears the OWC source tree is a treasure trove of useful information and routines that could be used to enhance ELKS, given OWC's massive support for 16-bit systems.

**Comment by rafael2k** [comment]
Now we have OW creating executables for Elks, I can more easily resume this by using the same watcom Makefile, but changing to ewcc and ewlink, and then change the network calls, right?

**Reply by ghaerr** [reply]
Yes, although the ewcc and ewlink may not exactly support (i.e. pass through) various options correctly, so you might possibly have to edit some of the options.

When you say "network calls", you mean the network system calls, like `connect`, `accept`, etc? Yes, they will need to be added. I've been meaning to do that, so when you get link errors with undefined system calls, let me know and I can add them to the ELKS OW C system call library, unless you'd like to.

**Reply by rafael2k** [reply]
Ha. Just realized that. I'll commit the new CFAGS to my forked repo (https://github.com/rafael2k/SSH2DOS). Btw, I see the libc compiles with owcc. Should I use owcc or wcc?

**Reply by ghaerr** [reply]
> I see the libc compiles with owcc. Should I use owcc or wcc?

We're using owcc now in `ewcc` and `ewlink`, because its options are easier to understand for Linux folks not used to OWC. Note that owcc is just a wrapper that calls wcc anyways. You probably ought to stay with using `ewcc` instead of owcc or wcc directly as there are a number of options that have to be included in order to generate the right code to run on ELKS. The same goes for header files - the OWC runtime DOS (or OS/2) libraries are generally not compatible with ELKS or the ELKS C library.

**Reply by rafael2k** [reply]
Got it. 
Indeed. I realized that header mixing was a problem, for example OW <io.h> has definitions for read() and write()! And of course many symbol clashes now I'm starting to port OW DOS software to Elks. Even more fun!
: )

**Comment by rafael2k** [comment]
Found some other option very interesting:
https://github.com/Oryx-Embedded/CycloneSSH
https://www.oryx-embedded.com/doc/files.html

**Comment by rafael2k** [comment]
Now a ssh server which is small enough I think could be ported to ELKS too:
https://github.com/janmojzis/tinyssh

**Reply by ghaerr** [reply]
Your list is getting longer!! 

TinySSH doesn't use dynamic memory allocation and code size is already almost 64K so I'd assume you'd need to use OWC in order to compile it in one of its models after figuring out how much data segment size is actually needed.

Did you ever take a pass at compiling SSH2DOS after we got OWC working with ELKS?

**Reply by rafael2k** [reply]
I started the work (with OWC, yes!), but it is some way to go. I would even consider another option, as SSH2DOS would need to have included newer algorithms to properly work.

**Comment by rafael2k** [comment]
I think this is long shot, but would still consider nice to have on ELKS, as many people really use ssh 99% of the time, and having ssh on ELKS would allow any ELKS box to be a daily-driver for some.
I still need to have a computer with networking up and running with ELKS.

**Comment by Oichkatzelesfrettschen** [comment]
Rafael - Have you looked using modern vintage hardware for your efforts,
such as looking into the Monotech NuXT (V20/8088+8087) or Pocket8086?

For the Monotech NuXT:

- Sale Link:
https://monotech.fwscart.com/product/nuxt-v2-0---microatx-turbo-xt---10mhz-832k-xt-ide-multi-io-svga
- Github Project Link:
https://monotech.fwscart.com/product/nuxt-v2-0---microatx-turbo-xt---10mhz-832k-xt-ide-multi-io-svga

Cool Add-Ons:

- Blasterboard (Soundblaster clone):
https://www.tindie.com/products/jrlab/blasterboard-a-complete-8-bit-isa-sound-card-2/
///////-Video Cards but the Monotech Includes a $15 SVGA add-on-\\\\\\\
- ISA CGA card:
https://www.tindie.com/products/cycle/ibm-compatible-color-graphic-adapter-isa-cga-card/
- ISA MDA or CGA New Old Stock adapters (Japanese market IBM MDA/CGA XT/AT
cards
- ISA SVGA SuperTrident TVGAi9000:
https://www.tindie.com/products/denjhang/isa-super-vga-trident-tvga9000i/

Here is the Pocket8086 which has a slew of external accessories and is used
by Youtubers as the "modern old school reference 16-bit reference platform"

https://www.tindie.com/products/cycle/pocket8086-retro-dos-computer-8086-10mhz-clear/

-----

There seems to be a lot of interest on various discords, on reddit, on IRC
about ELKs in general for retro-modern hardware as well as vintage-proper
hardware to have a machine to SSH into an AI or production machine, etc. as
well as have a convenient local OS.

**Comment by rafael2k** [comment]
Thanks @Oichkatzelesfrettschen! I'm more inclined to get a Pocket8086, but then I'll also need a NIC, or I just connect it to my main computer using a serial null modem cable. The NuXT is also very cool, but a bit pricey compared to the Pocket8086. Hopefully next month I'll take a decision on which modern vintage hardware to go.

**Comment by ghaerr** [comment]
Thanks for the updated list of modern vintage hardware @Oichkatzelesfrettschen!

The Pocket 8086 w/640x480 color VGA sounds pretty neat for sure. I see their advert shows it running Windows 3.0. I wonder how fast that is?

@rafael2k, ELKS SLIP works pretty well and with the special serial interrupt handler turned on via CONFIG_FAST_IRQ4 has allowed transfer speeds of 38.4K without UART receive overruns, although a hardware FIFO always helps. That's one of the reasons serial input queues are still set at 1K bytes.

**Reply by rafael2k** [reply]
> @rafael2k, ELKS SLIP works pretty well and with the special serial interrupt handler turned on via CONFIG_FAST_IRQ4 has allowed transfer speeds of 38.4K without UART receive overruns, although a hardware FIFO always helps. That's one of the reasons serial input queues are still set at 1K bytes.

Thanks for the information @ghaerr! Can the tcp/ip stack attach to the SLIP driver without problems? I was also thinking about porting Taylor's UUCP, which could be interesting for copying files in and out, and no encryption overhead.

**Comment by ghaerr** [comment]
> Can the tcp/ip stack attach to the SLIP driver without problems? 

`net start slip 9600 /dev/ttyS1`, then increase the baud rate and/or add CONFIG_FAST_IRQ4 (or IRQ3 for /dev/ttyS1) until you start seeing problems.

**Comment by toncho11** [comment]
https://github.com/bbenchoff/MacSSL

**Comment by rafael2k** [comment]
This could be good for https support.

**Comment by parabyte** [comment]
toncho thanks for the suggestion, not tried to the lib yet against a bin
<img width="528" height="82" alt="Screenshot_2025-11-02_18-24-06" src="https://github.com/user-attachments/assets/b1938a7d-f795-4a5f-832f-3efc2a23323b" />
but got macssl working with the elks toolchain, who knows if it works,i have not tried it yet in elks  will put source on github either today or tomorrow

**Reply by toncho11** [reply]
Wow! Very impressive if you managed to compile it on ELKS! 

Here is a very minimal code for a future ssh client.  It has to be used against a TLS 1.1. It assumes: MacSSL.h / MacSSL.lib provide the usual MbedTLS-style API: SSL_library_init(), SSL_CTX_new(), SSL_new(), SSL_connect(), SSL_read(), SSL_write(), SSL_shutdown(), SSL_free(), SSL_CTX_free()

```
/* ssh_minimal.c
 * Minimal SSH client skeleton for ELKS using MacSSL
 * Connects to host:22, initializes MacSSL library
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <conio.h>
#include <dos.h>

#include "MacSSL.h"   // Assuming MacSSL headers are available

#define SERVER "127.0.0.1"
#define PORT 22

int main(void) {
    int sock;
    SSL_CTX *ctx;
    SSL *ssl;

    printf("ELKS SSH minimal client using MacSSL\n");

    /* 1. Initialize MacSSL library */
    if (SSL_library_init() != 1) {
        printf("SSL library initialization failed\n");
        return 1;
    }
    SSL_load_error_strings();
    ctx = SSL_CTX_new(TLS_client_method());
    if (!ctx) {
        printf("Failed to create SSL context\n");
        return 1;
    }

    /* 2. Create TCP connection */
    sock = open_tcp_connection(SERVER, PORT); // you must implement open_tcp_connection() for ELKS
    if (sock < 0) {
        printf("Failed to open TCP connection\n");
        SSL_CTX_free(ctx);
        return 1;
    }

    /* 3. Create SSL object and attach to socket */
    ssl = SSL_new(ctx);
    SSL_set_fd(ssl, sock);

    /* 4. Perform SSL handshake */
    if (SSL_connect(ssl) <= 0) {
        printf("SSL handshake failed\n");
        SSL_free(ssl);
        SSL_CTX_free(ctx);
        close(sock);
        return 1;
    }

    printf("SSL handshake completed\n");

    /* 5. Send SSH protocol version string (example) */
    const char *ssh_hello = "SSH-2.0-ELKS-MacSSL\r\n";
    SSL_write(ssl, ssh_hello, strlen(ssh_hello));

    /* 6. Read server response */
    char buffer[256];
    int n = SSL_read(ssl, buffer, sizeof(buffer)-1);
    if (n > 0) {
        buffer[n] = '\0';
        printf("Server response: %s\n", buffer);
    }

    /* 7. Cleanup */
    SSL_shutdown(ssl);
    SSL_free(ssl);
    SSL_CTX_free(ctx);
    close(sock);

    printf("Connection closed\n");
    return 0;
}

/* Stub for TCP connection on ELKS - you will need to implement using ELKS sockets */
int open_tcp_connection(const char *host, int port) {
    /* Minimal ELKS TCP connect code goes here */
    /* Return socket fd or -1 on error */
    return -1; // placeholder
}

```

**Comment by toncho11** [comment]
Here is an improved ssh client using MacSSL. It should use large memory model with OWC.

```
/*
 * Minimal bi-directional "SSH-like" shell client for ELKS using MacSSL
 * 16-bit ELKS, compiled with Open Watcom C
 *
 * Features:
 * - Connects to a server over TCP
 * - Uses MacSSL (TLS library) for encrypted transport
 * - Sends password once (simple authentication)
 * - Receives text from server and prints it
 * - Sends user-typed commands to the server
 * - Detects if server requires TLS >1.1 or refuses all proposed cipher suites
 *
 * Limitations:
 * - MacSSL supports only TLS 1.1 or lower
 * - Ciphers: RSA + AES-CBC + SHA only
 * - Minimal console handling: no history, no editing
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <conio.h>
#include <dos.h>
#include "MacSSL.h"

#define SERVER "192.168.1.100"
#define PORT 22
#define PASSWORD "secret123"

int open_tcp_connection(const char *host, int port);

int main(void) {
    int sock;
    SSL_CTX *ctx;
    SSL *ssl;
    int ret;

    printf("ELKS minimal shell client (send & receive)\n");

    /* Initialize MacSSL */
    if (SSL_library_init() != 1) {
        printf("SSL initialization failed\n");
        return 1;
    }
    SSL_load_error_strings();

    ctx = SSL_CTX_new(TLS_client_method());
    if (!ctx) {
        printf("SSL_CTX_new failed\n");
        return 1;
    }

    sock = open_tcp_connection(SERVER, PORT);
    if (sock < 0) {
        printf("TCP connect failed\n");
        SSL_CTX_free(ctx);
        return 1;
    }

    ssl = SSL_new(ctx);
    SSL_set_fd(ssl, sock);

    /* Perform SSL handshake */
    ret = SSL_connect(ssl);
    if (ret <= 0) {
        int err = SSL_get_error(ssl, ret);

        /* TLS version mismatch */
        if (err == SSL_ERROR_PROTOCOL_VERSION) {
            printf("SSL handshake failed: server requires TLS 1.2 or higher\n");
        }
        /* Cipher mismatch or handshake failure otherwise */
        else {
            printf("SSL handshake failed: server refused all proposed cipher suites or other handshake error (error code %d)\n", err);
        }

        SSL_free(ssl);
        SSL_CTX_free(ctx);
        close(sock);
        return 1;
    }

    /* Handshake succeeded - print negotiated cipher */
    printf("SSL handshake successful (TLS 1.1)\n");
    printf("Using cipher: %s\n", SSL_get_cipher(ssl));

    /* Send password for authentication */
    SSL_write(ssl, PASSWORD, strlen(PASSWORD));
    SSL_write(ssl, "\n", 1);

    /* Main send & receive loop */
    char ch;
    char input_buffer[128];
    int i;

    while (1) {
        /* Receive server output */
        while (SSL_read(ssl, &ch, 1) == 1) {
            putchar(ch);
        }

        /* Read user input and send */
        printf("> ");
        if (fgets(input_buffer, sizeof(input_buffer), stdin) == NULL)
            break;

        i = strlen(input_buffer);
        if (i > 0 && input_buffer[i - 1] == '\n')
            input_buffer[i - 1] = '\0';

        SSL_write(ssl, input_buffer, strlen(input_buffer));
        SSL_write(ssl, "\n", 1);
    }

    /* Cleanup */
    SSL_shutdown(ssl);
    SSL_free(ssl);
    SSL_CTX_free(ctx);
    close(sock);

    printf("\nConnection closed\n");
    return 0;
}

/* Implement this for ELKS 16-bit TCP sockets */
int open_tcp_connection(const char *host, int port) {
    /* Placeholder - return -1 for now */
    return -1;
}
```

**Reply by toncho11** [reply]
And the open_tcp_connection(), using large memory model.

```
/*
 * open_tcp_connection()
 * --------------------
 * Minimal TCP connection function for ELKS 16-bit
 * Returns a connected socket descriptor or -1 on error.
 *
 * Parameters:
 *  host - string containing IPv4 address (e.g., "192.168.1.100")
 *  port - port number in host byte order
 *
 * Notes:
 *  - Uses ELKS BSD-style socket API
 *  - Compatible with Open Watcom 16-bit large memory model (-ml)
 *  - Must include <sys/socket.h> and <netinet/in.h> from ELKS
 */

#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>    // for close()
#include <string.h>
#include <stdio.h>

int open_tcp_connection(const char *host, int port) {
    int sockfd;
    struct sockaddr_in addr;

    /* 1. Create socket */
    sockfd = socket(AF_INET, SOCK_STREAM, 0);
    if (sockfd < 0) {
        printf("socket() failed\n");
        return -1;
    }

    /* 2. Initialize server address */
    memset(&addr, 0, sizeof(addr));
    addr.sin_family = AF_INET;
    addr.sin_port = htons(port);

    /* Convert host string to binary address */
    addr.sin_addr.s_addr = inet_addr(host);
    if (addr.sin_addr.s_addr == INADDR_NONE) {
        printf("Invalid host address: %s\n", host);
        close(sockfd);
        return -1;
    }

    /* 3. Connect to server */
    if (connect(sockfd, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
        printf("connect() failed\n");
        close(sockfd);
        return -1;
    }

    /* 4. Return connected socket */
    return sockfd;
}
```

**Comment by toncho11** [comment]
Actually your port must follow the segmented memory model of x86 in order to work correctly.

**Comment by parabyte** [comment]


**Reply by parabyte** [reply]
cleaned up the code a little today, instead of a kludge port. you may take control of this lib if you wish toncho?

ssl lib's are nearly outside of my comfort zone.

**Comment by rafael2k** [comment]
This is pretty cool! Do you have some sample application to use the library?

ps: just saw @toncho11 samples. Do you know if the functions are enough to run "plain" ssh with some wiring, may be using https://github.com/libssh2/libssh2

One idea if more dynamic memory is needed is to use the ELKS fmemalloc and friends: https://github.com/ghaerr/8086-toolchain/blob/dev86/compiler/mem.c

**Reply by parabyte** [reply]
i will create a new thread about this, sorry i should not of put this lib here

The current MacSSL shim only gives you a TLS 1.2 client; SSH needs its own key exchange and packet machinery, so you’d still need a full SSH stack libssh2 can target this lib, but would require some hefty figuring out!

so for example TLS 1.2 applies to protocols like HTTPS, FTPS, SMTPS, 

but its a step in the direction for elks ssh. 

I have only managed to get tls 1.2 working
urgh unfortunately im away for a month long family trip, i do not access to a computer now, only smartphone, much appreciated if you could report your findings, the sample code checks out against all relavent libs, im going to guess its going to segfault, iv literally not tested the code on a running system, just been bored staying away, had access to a computer.

**Reply by rafael2k** [reply]
Thanks @parabyte. Cool, having https and others which need tls 1.2 is pretty nice. We could try integrating to httpget.

**Reply by parabyte** [reply]
iv not looked at the code at all, but i wonder how much ram regular wget consumes!?

**Reply by parabyte** [reply]
iv given this its own thread now

**Reply by rafael2k** [reply]
we have urlget:
https://github.com/ghaerr/elks/tree/master/elkscmd/inet/urlget

**Comment by parabyte** [comment]
wget is  HUGE , to support paged 64k segments with large memmory model the rewriting of wget would be wild

**Comment by parabyte** [comment]
if your going to look around for ssh libs or ssh clients best option would be to look at embedded devices and projects for them. like arduino, esp32, but the lib needs to be platform neutral and not use floating point

**Reply by parabyte** [reply]
a lib specifically designed for a specific embeded device will have specific code for that specific device. makes it unfeasible to port the code, unless the developer knows that device sdk very well

**Comment by toncho11** [comment]
Here is our urlget reworked to use MacSSL:

[urlget.txt](https://github.com/user-attachments/files/23314181/urlget.txt)

**Comment by parabyte** [comment]
unfortuante i do not have access to a workstation, only this smartphone if anyone report there findings be very interesting
